traversal/iterator/GraphIterator.java

```diff
@@ -296,30 +298,22 @@ private boolean isClosure(ProcedureEdge<?, ?> edge, Vertex<?, ?> fromVertex, Ver
     }
 
     private boolean backTrack(int pos) {
-        backTrackCleanUp(pos);
+        popScope(pos);
         return computeNext(pos - 1);
     }
 
-    private void backTrackCleanUp(int pos) {
+    private void popScope(int pos) {
         ProcedureEdge<?, ?> edge = procedure.edge(pos);
-        Identifier toId = edge.to().id();
-        if (roles.containsKey(toId)) {
-            if (edge.isRolePlayer()) removeScopedRole(scopes.get(edge.asRolePlayer().scope()), edge);
-            else if (toId.isScoped()) removeScopedRole(scopes.get(toId.asScoped().scope()), edge);
-        }
-    }
-
-    private void removeScopedRole(Scopes.Scoped scoped, ProcedureEdge<?, ?> edge) {
-        ThingVertex previousRole = roles.remove(edge.to().id());
-        if (previousRole != null) scoped.remove(previousRole, edge.order());
-    }
-
-    private void replaceScopedRole(Scopes.Scoped scoped, ProcedureEdge<?, ?> edge, ThingVertex newRole) {
-        ThingVertex previousRole = roles.remove(edge.to().id());
-        if (previousRole != null) {
-            scoped.replaceRole(previousRole, newRole);
-        } else {
-            scoped.record(newRole, edge.order());
+        if (edge.to().id().isScoped()) {
+            Identifier.Variable scope = edge.to().id().asScoped().scope();
+            if (scopes.get(scope).orderVisited(pos)) scopes.get(scope).popLast();
+        } else if (edge.from().id().isScoped()) {
+            Identifier.Variable scope = edge.from().id().asScoped().scope();
+            assert scopes.get(scope).orderVisited(pos);
+            scopes.get(scope).popLast();
+        } else if (edge.isRolePlayer()) {
+            Identifier.Variable scope = edge.asRolePlayer().scope();
+            if (scopes.get(scope).orderVisited(pos)) scopes.get(scope).popLast();
```

Comment:
backtracking mirrors the `branch()` flow, where we scope roles depending on whether it is the `to()`, `from()` or shortcut edge that is scoped.